lib/media/segment_index.js

```diff
@@ -129,9 +129,24 @@ shaka.media.SegmentIndex.prototype.merge = function(references) {
       shaka.log.warning('Refusing to rewrite original references on update!');
       j++;
     } else {
-      // Drop the new reference if there's an old reference with the same
-      // time.
-      newReferences.push(r1);
+      // when period is changed, fitSegmentReference will
+      // expand last segment to the start of the next the period
+      // so, it is valid to have end time updated to the
+      // last segment reference in a period
+      if (r1.endTime != r2.endTime)
+      {
+        goog.asserts.assert(r2.endTime - r1.endTime <=
+            shaka.dash.MpdUtils.GAP_OVERLAP_TOLERANCE_SECONDS &&
+            i == this.references_.length - 1 &&
+            j == references.length - 1,
+            'This should be un update of the last segment in a period');
+        newReferences.push(r2);
+      }
+      else {
```

Comment:
Please move else to the previous line to match the style in the rest of the project:

} else {